File: pythonLeetcode/28-实现strStr.py
"""
实现 strStr() 函数。

给定一个 haystack 字符串和一个 needle 字符串，在 haystack 字符串中找出 needle 字符串出现的第一个位置 (从0开始)。如果不存在，则返回  -1。

示例 1:

输入: haystack = "hello", needle = "ll"
输出: 2
示例 2:

输入: haystack = "aaaaa", needle = "bba"
输出: -1
"""

class Solution:
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if len(haystack) < len(needle):
            return -1
        if needle == None or needle == "":
            return 0
        needLength = len(needle)

        lastIndex = len(haystack) - needLength - 1
        j = 0
        while j <= lastIndex:
            if haystack[j:j+needLength] == needle:
                return j
            else:
                j = j+1
        return -1

File: LoveSwiftLeetcode/28.实现-str-str.swift
/*
 * @lc app=leetcode.cn id=28 lang=swift
 *
 * [28] 实现 strStr()
 */

// @lc code=start
class Solution {
    func strStr(_ haystack: String, _ needle: String) -> Int {
        if haystack.count < needle.count { return -1 }
        if haystack.count > 0 && needle.count == 0 { return 0 }
        let needleCnt = needle.count
        let lastIndex = haystack.count - needleCnt
        var start = 0
        while start <= lastIndex {
            let subRange = haystack.index(haystack.startIndex, offsetBy: start)..<haystack.endIndex
            if haystack[subRange].hasPrefix(needle) {
                return start
            }
            start += 1
        }
        return -1
    }
}
// @lc code=end

